### ISSUE TITLE
MA0182 false positive with factory

### ISSUE THREAD

**Issue by Piedone** [meziantou/Meziantou.Analyzer#970]
### Version of the Meziantou.Analyzer NuGet package

2.0.278

### Rule Identifier

MA0182

### Target Framework

net8.0

### C# Language version

C# 12

### Description

MA0182 shows a false positive violation when the class uses the factory pattern. See the repro, but below are some further real-life examples.

E.g. I get the below error:

```
Error: /home/runner/work/Open-Source-Orchard-Core-Extensions/Open-Source-Orchard-Core-Extensions/test/Lombiq.UITestingToolbox/Lombiq.Tests.UI/Services/GitHub/GitHubActionsGroupingTestOutputHelper.cs(7,23): error MA0182: Internal type 'GitHubActionsGroupingTestOutputHelper' is apparently never used. If so, remove it from the assembly. If this type is intended to contain only static members, make it 'static'. (https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/Rules/MA0182.md) [/home/runner/work/Open-Source-Orchard-Core-Extensions/Open-Source-Orchard-Core-Extensions/test/Lombiq.UITestingToolbox/Lombiq.Tests.UI/Lombiq.Tests.UI.csproj]
```

This is for this file: https://github.com/Lombiq/UI-Testing-Toolbox/blob/135cb1f3606959fd6d95133753953853c0ded570/Lombiq.Tests.UI/Services/GitHub/GitHubActionsGroupingTestOutputHelper.cs. As you can see, while the class is `internal` and its ctor `private`, its `CreateDecorator` method serves as a factory method. That is then used here: https://github.com/Lombiq/UI-Testing-Toolbox/blob/135cb1f3606959fd6d95133753953853c0ded570/Lombiq.Tests.UI/UITestBase.cs#L31

It's the same issue with this class: https://github.com/Lombiq/Orchard-Core-API-Client/blob/2a76f21b4d3a4202c58cbd75e771d3415c39a525/Lombiq.OrchardCoreApiClient/ConfigurableCertificateValidatingHttpClientHandler.cs.

```
Error: /home/runner/work/Open-Source-Orchard-Core-Extensions/Open-Source-Orchard-Core-Extensions/src/Libraries/Lombiq.OrchardCoreApiClient/Lombiq.OrchardCoreApiClient/ConfigurableCertificateValidatingHttpClientHandler.cs(23,23): error MA0182: Internal type 'ConfigurableCertificateValidatingHttpClientHandler' is apparently never used. If so, remove it from the assembly. If this type is intended to contain only static members, make it 'static'. (https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/Rules/MA0182.md) [/home/runner/work/Open-Source-Orchard-Core-Extensions/Open-Source-Orchard-Core-Extensions/src/Libraries/Lombiq.OrchardCoreApiClient/Lombiq.OrchardCoreApiClient/Lombiq.OrchardCoreApiClient.csproj]
```

This also implements the factory pattern, with the ctor being `private`, and `CreateClient` serving as the factory method. It's used here: https://github.com/Lombiq/Orchard-Core-API-Client/blob/2a76f21b4d3a4202c58cbd75e771d3415c39a525/Lombiq.OrchardCoreApiClient/ApiClient.cs#L28

### Reproduction Steps

I can't seem to be able to reproduce this locally for some reason, it only happens [in CI](https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions/actions/runs/21045354474/job/60518045836?pr=1214). Even if I set the exact same SDK with `global.json`, no violation locally. So, this might depend on the environment in some way (or be a Linux/Windows difference).

```c#
internal sealed class BugDemo
{
    private BugDemo()
    {
    }

    public static BugDemo Create() => new();
}

// ...

var x = BugDemo.Create();

```

This produces:

```
error MA0182: Internal type 'BugDemo' is apparently never used. If so, remove it from the assembly. If this type is intended to contain only static members, make it 'static'. (https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/Rules/MA0182.md)
```

### Other information

This started for me when updating to 2.0.278 from 2.0.270.

**Comment by meziantou** [comment]
The latest version (2.0.279) better handles method invocations which should solve this issue. It also contains fixes for extension (C# 14) and entrypoint.

**Comment by Piedone** [comment]
That was quick, thank you! It indeed fixed the issue.
